src/main/ruby/truffleruby/core/time.rb

```diff
@@ -336,7 +336,13 @@ def at(sec, usec=undefined, unit=undefined, **kwargs)
           raise ArgumentError, "unexpected unit: #{unit}"
         end
 
-      usec = 0 if Primitive.undefined?(usec)
+      if Primitive.undefined?(usec) && Primitive.object_kind_of?(sec, Rational)
+        sec, fractional_seconds = sec.divmod(1)
+        micro_fractional_seconds = fractional_seconds * 1_000_000
+        usec = micro_fractional_seconds.to_i
+      elsif Primitive.undefined?(usec)
+        usec = 0 if Primitive.undefined?(usec)
+      end
```

Comment:
This is an improvement for microseconds but still loses precision with nanoseconds:

```
>> t1 = Time.at(Rational(1_486_570_508_539_759, 1_000_000_000))
=> 1970-01-18 05:56:10.508539 +0100
>> t2 = Time.at(1_486_570, 508_539_759, :nsec)
=> 1970-01-18 05:56:10.508539759 +0100
>> t1.usec == t2.usec
=> true
>> t1.nsec == t2.nsec
=> false
>> t1.nsec
=> 508539000
>> t2.nsec
=> 508539759
```

In principle that could be a problem for a separate PR, but it’s broken (at least) [a `Time#-` spec](https://github.com/oracle/truffleruby/runs/2056645766#step:7:34).

Reply:
Fixed in latest version of the commit!